### PR TITLE
Cache total size in file based source

### PR DESCRIPTION
**Pull request by Abacn**
In some use case it is found excessive match request (up to 6 times) during splitting file based source (e.g. in BigQueryTableSource) and causing split request taking lots of time when there are lots of files.

This value can be cached at minimum cost and avoid actual match api call.

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
Some sources e.g. BigQueryTableSource already implemented cache:

https://github.com/apache/beam/blob/7caadb216df0794ad5c9a73401b51696104fd99f/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableSource.java#L71-L73

**Comment by Abacn** [comment]
R: @johnjcasey

**Comment by Abacn** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by Abacn** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by Abacn** [comment]
Run Java PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java [hunk 1]
```diff
@@ -91,6 +94,7 @@ protected FileBasedSource(
     this.mode = Mode.FILEPATTERN;
     this.emptyMatchTreatment = emptyMatchTreatment;
     this.fileOrPatternSpec = fileOrPatternSpec;
+    this.filesSizeBytes = new AtomicReference<>();
```
Comment (johnjcasey) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java:97]: Is this a file resource level cache?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java [hunk 2]
```diff
@@ -73,6 +74,8 @@
   private MatchResult.@Nullable Metadata singleFileMetadata;
   private final Mode mode;
 
+  private final AtomicReference<@Nullable Long> filesSizeBytes;
```
Comment (johnjcasey) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java:77]: Can we add a test case for this?

**Review by johnjcasey — CHANGES_REQUESTED** [review]

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java (hunk 1 above)
Comment (Abacn) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java:97]: No, it's just a Long value

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java (hunk 2 above)
Comment (Abacn) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java:77]: Done

**Pull request merged**